# **03 - Machine Learning**

## Objectives

- Prepare the cleaned retail data for machine learning.
- Define an appropriate machine learning problem based on the exploratory analysis.
- Select and justify a suitable algorithm.
- Train and evaluate the model using appropriate performance metrics.
- Interpret the model results and assess its usefulness for the business.
- Identify limitations and potential improvements.

## Inputs

- Cleaned Online Retail Transaction dataset produced during the ETL process and explored during EDA.

## Outputs

- A trained and evaluated machine learning model.
- Model performance metrics and visualisations.
- Interpretation of the model results and their business relevance.

## Additional Comments

- The modelling approach will be selected based on the business problem and characteristics identified during exploratory analysis.

---

# Change working directory

In [1]:
import os

current_dir = os.getcwd()
print("Current directory:", current_dir)

if os.path.basename(current_dir) == "jupyter_notebooks":
    os.chdir(os.path.dirname(current_dir))
elif os.path.basename(current_dir) == "vscode-projects":
    os.chdir("Online Retail Transaction Analysis")

current_dir = os.getcwd()
print("Working directory set to:", current_dir)

Current directory: /Users/sahraosman/Documents/vscode-projects/Online Retail Transaction Analysis/jupyter_notebooks
Working directory set to: /Users/sahraosman/Documents/vscode-projects/Online Retail Transaction Analysis


---

# Section 1 -  Import libraries

Here I will be importing the libraries that I will be using in this notebook.

In [2]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.preprocessing import StandardScaler
from sklearn.cluster import KMeans
from sklearn.metrics import silhouette_score

---

# Section 2 - Load the dataset

The cleaned Online Retail dataset produced during the ETL stage will be loaded for the machine learning analysis. I will first inspect the dataset to confirm that it has loaded correctly before preparing the data for modelling.

In [ ]:
df = pd.read_csv(
    "outputs/online_retail_cleaned.csv",
    dtype={"InvoiceNo": str},
    parse_dates=["InvoiceDate"],
)

In [4]:
df.head()

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country,Revenue
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,2010-12-01 08:26:00,2.55,17850,United Kingdom,15.30
1,536365,71053,WHITE METAL LANTERN,6,2010-12-01 08:26:00,3.39,17850,United Kingdom,20.34
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,2010-12-01 08:26:00,2.75,17850,United Kingdom,22.00
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,2010-12-01 08:26:00,3.39,17850,United Kingdom,20.34
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,2010-12-01 08:26:00,3.39,17850,United Kingdom,20.34


In [5]:
df.shape

(524878, 9)

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 524878 entries, 0 to 524877
Data columns (total 9 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   InvoiceNo    524878 non-null  object        
 1   StockCode    524878 non-null  object        
 2   Description  524878 non-null  object        
 3   Quantity     524878 non-null  int64         
 4   InvoiceDate  524878 non-null  datetime64[ns]
 5   UnitPrice    524878 non-null  float64       
 6   CustomerID   524878 non-null  int64         
 7   Country      524878 non-null  object        
 8   Revenue      524878 non-null  float64       
dtypes: datetime64[ns](1), float64(2), int64(2), object(4)
memory usage: 36.0+ MB


The cleaned dataset contains 524,878 rows and 9 columns, with no missing values. The data types have also loaded as expected, including `InvoiceDate` as a datetime.

This confirms that the cleaned dataset has loaded correctly and is ready to be prepared for the machine learning analysis.

---

# Section 3 - Define the Machine Learning Problem

The exploratory analysis identified differences in customer purchasing behaviour. Customers varied in how frequently they made purchases and in the total revenue they generated. The analysis also showed that high average transaction values could sometimes be influenced by a small number of unusually large purchases.

Based on these findings, customer segmentation was selected as the machine learning problem. The aim is to identify groups of customers with similar purchasing behaviours using Recency, Frequency and Monetary (RFM) measures.

As the dataset does not contain predefined customer segment labels, this is an unsupervised machine learning problem. K-Means clustering will be used to group customers based on similarities in their RFM characteristics.